dotnet.cmd

```diff
@@ -0,0 +1,2 @@
+@echo off
+powershell -ExecutionPolicy ByPass -NoProfile -command "& """%~dp0dotnet.ps1""" %*"
```

Comment:
I do not think we should be executing dotnet indirectly via powershell. These wrappers over wrappers just cause pain.

Reply:
The Arcade powershell script also locates the dotnet cli as well as downloading it if it isn't present. The previous version of dotnet.cmd also called into powershell to do the same thing, but in a less straight-forward manner.

Reply:
I am fine with calling powershell to download and locale dotnet.exe. I would like to see the actual invocation of dotnet.exe to be done directly from the .cmd/.sh files.

Reply:
I believe that the diagnosability is indirectly proportion to the max depth of the process tree. (My memories of `run` tool that added an extra process layer everywhere and made problems so much harder to troubleshoot are still too fresh.)

We are not introducing the extra process layer on Unix and invoking dotnet directly from the .sh script. Can we do it the same way on Windows too?

Reply:
Thanks for letting us know, I will explore other options without the indirection. On Unix this is fine as the arcade scripts can be `dot` sourced into the caller as the environment and the shell engine is the same. On Windows this is more tricky as their is no standard mechanism to share variables between cmd and powershell scripts besides stdin/stdout, registry, global vars, etc. I will probably find another way to do this.

Reply:
> I will probably find another way to do this.

After researching a bit I'm not less sure about this.

Reply:
What is wrong with the `for /f ...` used in the current script (https://github.com/dotnet/runtime/blob/master/src/coreclr/dotnet.cmd#L21)? That's one of the best way to do this, as far as I know.

Reply:
> What is wrong with the for /f ... used in the current script

That all error handling and stdout in the arcade script is swallowed. The arcade function is not responsible for returning the path to the SDK folder but also to add telemetry markers when being used in CI. There must be a better way to do this.

Reply:
@jkotas I addressed your concerns, the calling script now invokes the host. The powershell script now has the following flow:

1. Retrieve the SDK folder
2a. Call the host if the script was invoked in a normal way
2b. Call the dotnet.cmd script if the `Get-DotNetPath` function was invoked.

Similar for the dotnet.cmd script which detects if the SDK directory is passed in as argument or not.

Reply:
Oh I believe the issue with process tree additional depth is still present.

Reply:
OK one more iteration, I'm now using a temporary file to store the location similar to what @trylek tried initially. This seems to be the cleanest option.

Reply:
> That all error handling and stdout in the arcade script is swallowed

`for /f ... ` won't swallow error handling. If the command prints something to stderr, it gets passed through.

Are you saying that Arcade scripts print errors to stdout instead of stderr?

Reply:
As mentioned the Arcade scripts do pipeline logging as well, ie https://github.com/dotnet/runtime/blob/285911db951b2721c46f4e5d65bf1b419ae07c03/eng/common/tools.ps1#L157-L160.

> for /f ... won't swallow error handling. If the command prints something to stderr, it gets passed through.

It does, it swallows the default powershell exceptions, I just tried that locally when checking out my first commit in this branch. If I keep the `@echo off` I don't get any output. If I remove the `@echo off` I get the following output:

```
C:\git\runtime2>.\dotnet.cmd

C:\git\runtime2>setlocal

C:\git\runtime2>set Platform=

C:\git\runtime2>set DOTNET_MULTILEVEL_LOOKUP=0

C:\git\runtime2>set DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1

C:\git\runtime2>set PS_DOTNET_INSTALL_SCRIPT=". C:\git\runtime2\eng\common\tools.ps1; InitializeDotNetCli($true)"

C:\git\runtime2>set "PS_COMMAND=powershell -NoProfile -ExecutionPolicy unrestricted -Command ". C:\git\runtime2\eng\common\tools.ps1; InitializeDotNetCli($true)""

C:\git\runtime2>for /F "delims=" %l in ('powershell -NoProfile -ExecutionPolicy unrestricted -Command ". C:\git\runtime2\eng\common\tools.ps1; InitializeDotNetCli($true)"') do set "__dotnetDir=%l"

C:\git\runtime2>set "__dotnetDir=Invoke-WebRequest : The remote name could not be resolved: 'dot.net'"

C:\git\runtime2>set "__dotnetDir=At C:\git\runtime2\eng\common\tools.ps1:170 char:5"

C:\git\runtime2>set "__dotnetDir=+     Invoke-WebRequest "https://dot.net/$dotnetInstallScriptVersion/do ..."

C:\git\runtime2>set "__dotnetDir=+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"

C:\git\runtime2>set "__dotnetDir=    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebExc "

C:\git\runtime2>set "__dotnetDir=   eption"

C:\git\runtime2>set "__dotnetDir=    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand"

C:\git\runtime2>set "__dotnetDir= "

C:\git\runtime2>set "dotnetPath= \dotnet.exe"

C:\git\runtime2>call " \dotnet.exe"
'" \dotnet.exe"' is not recognized as an internal or external command,
operable program or batch file.

C:\git\runtime2>.\dotnet.cmd
```

Reply:
> It does, it swallows the default powershell exceptions

That's because of powershell seems to like to print exceptions by default to stdout, not to stderr.

If I create powershell script `oops.ps1` that prints an error to stderr:
```
[Console]::Error.WriteLine('oops')
```

And call it from a `test.bat` like this:

```
@echo  off
for /f "delims=" %%l in ('powershell -NoProfile -ExecutionPolicy unrestricted -Command .\oops.ps1') do set "__dotnetDir=%%l"
```

I do see the error when invoking the .bat file:

```
D:\repro>test.bat
oops
```